crd/src/util.rs

```diff
@@ -0,0 +1,674 @@
+
+    LabelSelector {
+        match_labels: Some(opa_pod_matchlabels),
+        ..Default::default()
+    }
+}
+
+/// Check in kubernetes, whether the OPA object referenced by `opa_name` and `opa_namespace`
+/// exists. If it exists the object will be returned.
+async fn check_opa_reference(
+    client: &Client,
+    opa_name: &str,
+    opa_namespace: &str,
+) -> OpaOperatorResult<OpenPolicyAgent> {
+    debug!(
+        "Checking OpaReference if [{}] exists in namespace [{}].",
+        opa_name, opa_namespace
+    );
+    let opa_cluster: OperatorResult<OpenPolicyAgent> =
+        client.get(opa_name, Some(opa_namespace)).await;
+
+    opa_cluster.map_err(|err| {
+        warn!(?err,
+                        "Referencing a OPA that does not exist (or some other error while fetching it): [{}/{}], we will requeue and check again",
+                        opa_namespace,
+                        opa_name
+                    );
+        OperatorFrameworkError {source: err}})
+}
+
+/// Builds the actual connection string after all necessary information has been retrieved.
+/// As best practice and to reduce network traffic and increase response time, if a node_name
+/// is provided and matches one of the pod node_name, this node_name is selected. Without a match
+/// a random node is returned. TODO: for now the node_names are sorted and the last node is returned.
+///
+/// # Arguments
+///
+/// * `opa_spec` - OpaSpec to retrieve different port configuration
+/// * `opa_pods` - Pods belonging to the OPA cluster
+/// * `opa_api` - The desired OPA endpoint like Data, Query, Policy etc.
+/// * `opa_api_protocol` - The desired OPA endpoint protocol like HTTP or HTTPS
+/// * `desired_node_name` - If node_name is provided we look for opa deployments on the same node name to improve response time
+///
+fn get_opa_connection_string_from_pods(
+    opa_spec: &OpaSpec,
+    opa_pods: &[Pod],
+    opa_api: &OpaApi,
+    opa_api_protocol: &OpaApiProtocol,
+    desired_node_name: Option<String>,
+) -> OpaOperatorResult<String> {
+    let mut server_and_port_list = Vec::new();
+
+    for pod in opa_pods {
+        let pod_name = match pod.metadata.name.as_ref() {
+            None => {
+                return Err(ObjectWithoutName {
+                    reference: ErrOpaPodWithoutName.to_string(),
+                })
+            }
+            Some(pod_name) => pod_name,
+        };
+
+        let node_name = match pod.spec.as_ref().and_then(|spec| spec.node_name.as_ref()) {
+            None => {
+                debug!("Pod [{:?}] is does not have node_name set, might not be scheduled yet, aborting.. ",
+                       pod_name);
+                return Err(PodWithoutHostname {
+                    pod: pod_name.clone(),
+                });
+            }
+            Some(node_name) => node_name,
+        };
+
+        let role_group = match pod
+            .metadata
+            .clone()
```

Comment:
The clone is dirty but I see why you're doing it.
Removing it would mean a nested match I believe.
Up to you. Fine for me as it is.